### PR TITLE
(VANAGON-57) inform on STDERR, not STDOUT

### PR DESCRIPTION
**Pull request by mckern**
I'm not in love with this approach, but it's not drastically different
than how Vanagon handled notification/informing users of state/status
previously. The only difference is now the output is explicit, and can
be more easily/predictably filtered. I think that this is a hold-over
bandaid until we've got a better logging/notification system in place
in Vanagon but it should address the concerns in VANAGON-57.

As is, I'm not wedded to this approach and I'd love to hear any ideas anyone else has.

### CODE REVIEW

File: lib/vanagon/component/source/http.rb [hunk 1]
```diff
@@ -78,7 +78,7 @@ def file
         #
         # @raise [RuntimeError] an exception is raised if the sum does not match the sum of the file
         def verify # rubocop:disable Metrics/AbcSize
-          puts "Verifying file: #{file} against sum: '#{sum}'"
+          $stderr.puts "Verifying file: #{file} against sum: '#{sum}'"
```
Comment (McdonaldSeanp) [inline lib/vanagon/component/source/http.rb]: This seems like it should go to stdout

File: lib/vanagon/component/source/http.rb [hunk 2]
```diff
@@ -92,7 +92,7 @@ def download(target_url, target_file = nil, headers = { "Accept-Encoding" => "id
           uri = URI.parse(target_url.to_s)
           target_file ||= File.basename(uri.path)
 
-          puts "Downloading file '#{target_file}' from url '#{target_url}'"
+          $stderr.puts "Downloading file '#{target_file}' from url '#{target_url}'"
```
Comment (McdonaldSeanp) [inline lib/vanagon/component/source/http.rb]: Also should go to stdout

File: lib/vanagon/driver.rb [hunk 3]
```diff
@@ -148,7 +148,7 @@ def render
         raise Vanagon::Error, "Project requires a version set, all is lost."
       end
 
-      puts "rendering Makefile"
+      $stderr.puts "rendering Makefile"
```
Comment (McdonaldSeanp) [inline lib/vanagon/driver.rb]: should go to stdout

File: lib/vanagon/driver.rb [hunk 4]
```diff
@@ -117,7 +117,7 @@ def run # rubocop:disable Metrics/AbcSize
 
       @engine.startup(@workdir)
 
-      puts "Target is #{@engine.target}"
+      $stderr.puts "Target is #{@engine.target}"
```
Comment (McdonaldSeanp) [inline lib/vanagon/driver.rb]: should go to stdout

File: lib/vanagon/driver.rb [hunk 5]
```diff
@@ -43,8 +43,8 @@ def filter_out_components(only_build)
       # flatten all the results in to one array and set project.components to that.
       @project.components = only_build.flat_map { |comp| @project.filter_component(comp) }.uniq
       if @verbose
-        puts "Only building:"
-        @project.components.each { |comp| puts comp.name }
+        $stderr.puts "Only building:"
```
Comment (McdonaldSeanp) [inline lib/vanagon/driver.rb]: should go to stdout

File: lib/vanagon/engine/ec2.rb [hunk 6]
```diff
@@ -57,17 +57,17 @@ def instance
       end
 
       def select_target
-        puts "Instance created id: #{instance.id}"
-        puts "Created instance waiting for status ok"
+        $stderr.puts "Instance created id: #{instance.id}"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/ec2.rb]: Should go to stdout

File: lib/vanagon/engine/ec2.rb [hunk 7]
```diff
@@ -57,17 +57,17 @@ def instance
...
+        $stderr.puts "Created instance waiting for status ok"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/ec2.rb]: should go to stdout

File: lib/vanagon/engine/ec2.rb [hunk 8]
```diff
@@ -57,17 +57,17 @@ def instance
...
         @ec2.wait_until(:instance_status_ok, instance_ids: [instance.id])
-        puts "Instance running"
+        $stderr.puts "Instance running"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/ec2.rb]: should go to stdout

File: lib/vanagon/engine/ec2.rb [hunk 9]
```diff
@@ -57,17 +57,17 @@ def instance
...
         @target = instance.private_ip_address
       rescue ::Aws::Waiters::Errors::WaiterFailed => error
         fail "Failed to wait for ec2 instance to start got error #{error}"
       end
 
       def teardown
-        puts "Destroying instance on AWS id: #{instance.id}"
+        $stderr.puts "Destroying instance on AWS id: #{instance.id}"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/ec2.rb]: should go to stdout

File: lib/vanagon/engine/hardware.rb [hunk 10]
```diff
@@ -23,7 +23,7 @@ def polling_lock(host)
         Vanagon::Driver.logger.info "Polling for a lock on #{host}."
         @lockman.polling_lock(host, VANAGON_LOCK_USER, "Vanagon automated lock")
         Vanagon::Driver.logger.info "Lock acquired on #{host}."
-        puts "Lock acquired on #{host} for #{VANAGON_LOCK_USER}."
+        $stderr.puts "Lock acquired on #{host} for #{VANAGON_LOCK_USER}."
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/hardware.rb]: should go to stdout

File: lib/vanagon/engine/hardware.rb [hunk 11]
```diff
@@ -33,7 +33,7 @@ def node_lock(hosts)
           Vanagon::Driver.logger.info "Attempting  to lock #{h}."
           if @lockman.lock(h, VANAGON_LOCK_USER, "Vanagon automated lock")
             Vanagon::Driver.logger.info "Lock acquired on #{h}."
-            puts "Lock acquired on #{h} for #{VANAGON_LOCK_USER}."
+            $stderr.puts "Lock acquired on #{h} for #{VANAGON_LOCK_USER}."
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/hardware.rb]: should go to stdout

File: lib/vanagon/engine/hardware.rb [hunk 12]
```diff
@@ -45,7 +45,7 @@ def node_lock(hosts)
       # complete. In this case, we'll attempt to unlock the hardware
       def teardown
         Vanagon::Driver.logger.info "Removing lock on #{@target}."
-        puts "Removing lock on #{@target}."
+        $stderr.puts "Removing lock on #{@target}."
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/hardware.rb]: should go to stdout

File: lib/vanagon/engine/pooler.rb [hunk 13]
```diff
@@ -56,7 +56,7 @@ def read_vanagon_token(path = "~/.vanagon-token")
         absolute_path = File.expand_path(path)
         return nil unless File.exist?(absolute_path)
 
-        puts "Reading vmpooler token from: #{path}"
+        $stderr.puts "Reading vmpooler token from: #{path}"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/pooler.rb]: should go to stdout

File: lib/vanagon/engine/pooler.rb [hunk 14]
```diff
@@ -70,7 +70,7 @@ def read_vmfloaty_token(path = "~/.vmfloaty.yml")
...
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/pooler.rb]: should go to stdout

File: lib/vanagon/engine/pooler.rb [hunk 15]
```diff
@@ -119,7 +119,7 @@ def teardown
         )
         if response and response["ok"]
           Vanagon::Driver.logger.info "#{@target} has been destroyed"
-          puts "#{@target} has been destroyed"
+          $stderr.puts "#{@target} has been destroyed"
```
Comment (McdonaldSeanp) [inline lib/vanagon/engine/pooler.rb]: should go to stdout

File: lib/vanagon/optparse.rb [hunk 16]
```diff
@@ -28,7 +28,7 @@ def initialize(banner, options = [])
         end
 
         opts.on('-h', '--help', 'Display help') do
-          puts opts
+          $stdout.puts opts
           exit 1
```
Comment (McdonaldSeanp) [inline lib/vanagon/optparse.rb]: unrelated, but why would displaying help return 1 and not 0?

File: lib/vanagon/project/dsl.rb [hunk 17]
```diff
@@ -228,7 +228,7 @@ def identifier(ident)
       #
       # @param name [String] name of component to add. must be present in configdir/components and named $name.rb currently
       def component(name)
-        puts "Loading #{name}" if @project.settings[:verbose]
+        $stderr.puts "Loading #{name}" if @project.settings[:verbose]
```
Comment (McdonaldSeanp) [inline lib/vanagon/project/dsl.rb]: should go to stdout

File: lib/vanagon/engine/pooler.rb (hunk 13 above)
Comment (Magisus) [inline lib/vanagon/engine/pooler.rb]: This is exactly the message that is screwing us up in CI by being printed to stdout.

**Comment by Magisus** [comment]
@McdonaldSeanp this is in response to https://github.com/puppetlabs/ci-job-configs/pull/2649. Not sure if you were following that conversation. Do you have another approach?

File: lib/vanagon/optparse.rb (hunk 16 above)
Comment (mckern) [inline lib/vanagon/optparse.rb]: You know, I actually wonder who did that too. Asking for help and getting help means it did what you asked for. I'll fix that in this PR.

**Comment by mckern** [comment]
@McdonaldSeanp So. Let's talk about STDOUT/STDERR paradigms. Errors go to STDERR, everything else goes to STDOUT, right? BUT! Because we've only got two default output streams (technically you have a ton more, but `&0`, `&1`, and `&2` are the only ones you reliably have). So, why would we send notices to STDERR and not STDOUT? Because those notices are not actually required to understand state -- they're strictly informative. Example: curl.

Here's a `curl` command:

```
mckern@flexo Desktop $ curl --output --progress-bar --location https://raw.githubusercontent.com/rbenv/ruby-build/9b5123c7fc0766b0dbd16eb5a11167e487d58db6/share/ruby-build/2.2.7
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   386  100   386    0     0   4769      0 --:--:-- --:--:-- --:--:--  4825
mckern@flexo Desktop $
```

Let's redirect STDOUT:

```
mckern@flexo Desktop $ curl --output --progress-bar --location https://raw.githubusercontent.com/rbenv/ruby-build/9b5123c7fc0766b0dbd16eb5a11167e487d58db6/share/ruby-build/2.2.7 >/dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   386  100   386    0     0   2507      0 --:--:-- --:--:-- --:--:--  2522
mckern@flexo Desktop $
```

Huh. What changed? Nothing meaningful, because the informative but not-essential output was written to STDERR in light of a lack of STDINFO or STDNICETOKNOW or STDOHTHATMAKESSENSE. I suspect that we'll often end up doing the same thing in Vanagon, and that's why so many of these informative-but-not-essential messages were redirected to STDERR instead of STDOUT.

So, with that background, where do you think some of those messages should go?

**Comment by McdonaldSeanp** [comment]
@mckern yeah, I get why we would put these things in stderr. It bothers me to push everything to stderr since most of it is just verbose stdout but we currently don't have a great alternative. 

Is there work ticketed to get more powerful logging in vanagon? Like the ability to specify things will go to stdout when -v/--verbose are specified and either a logfile or nowhere otherwise?

**Comment by McdonaldSeanp** [comment]
What I would like to see eventually is the ability to do something like: 

```
Vanagon::Driver.logger.verbose "something infomative that could go to stdout but doesn't /need/ to"
```

**Comment by Magisus** [comment]
Do we not have any standard logging utility for Ruby?

**Comment by mckern** [comment]
rofl-copter

**Comment by underscorgan** [comment]
This seems reasonable I think. Mostly worried about what this might break. Have you built anything with this?

**Comment by shrug** [comment]
Will merge after Vanagon 0.11.3 release... hopefully today

**Pull request merged**